dvc/cache/base.py

```diff
@@ -89,7 +89,27 @@ def load_dir_cache(self, hash_info):
 
         return DirInfo.from_list(d)
 
-    def changed(self, path_info, hash_info):
+    def _get_filtered_hash_info(self, filter_info, hash_info, path_info):
+        dir_info = self.get_dir_cache(hash_info)
+        hash_key = filter_info.relative_to(path_info).parts
+
+        # Check whether it is a file that exists on the trie
+        if hash_key in dir_info.trie:
+            return filter_info, dir_info.trie[hash_key]
+
+        depth = len(hash_key)
+        filtered_dir_info = DirInfo()
+        try:
+            for key, value in dir_info.trie.items(hash_key):
+                filtered_dir_info.trie[key[depth:]] = value
+        except KeyError:
+            # The directory doesn't exist on the trie, fallback to the original
+            return path_info, hash_info
+        else:
+            filtered_hash_info, _ = self._get_dir_info_hash(filtered_dir_info)
+            return filter_info, filtered_hash_info
```

Comment:
Note how the first return value is either `filter_info` or `path_info`(just a binary switch). There should be a nicer way to handle this.

Reply:
They both are `PathInfo`s, and I don't know what else I can return there. Maybe just raise an exception when we can't filter (just like a signal) and then use catch it right away?